livebooks/vax_example.livemd

```diff
@@ -0,0 +1,444 @@
+  |> Repo.insert()
+
+{:ok, %{reactions: 5}} =
+  post_base
+  |> Vax.Types.Counter.cast_increment(:reactions, 3)
+  |> Repo.update()
+
+# updating stale, incs get merged
+{:ok, %{reactions: 8}} =
+  post_base
+  |> Ecto.Changeset.cast(%{reactions: 5}, [:reactions])
+  |> Repo.update()
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+{:ok,
+ %VaxExample.PostWithCrdts{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "post_with_crdts">,
+   content: nil,
+   id: "9f36d962-242f-4334-961e-7e9eb2e52389",
+   reactions: 8,
+   tags: #MapSet<[]>,
+   title: "Concurrency safe counter and set"
+ }}
+```
+
+```elixir
+{:ok, post_base} =
+  %VaxExample.PostWithCrdts{title: "Concurrency safe counter and set"}
+  |> Repo.insert()
+
+expected_tags = MapSet.new(["cool"])
+
+{:ok, %{tags: ^expected_tags}} =
+  post_base
+  |> Ecto.Changeset.change(%{tags: ["cool"]})
+  |> Repo.update()
+
+# updating stale, incs get merged
+expected_tags = MapSet.new(["cool", "new"])
+
+{:ok, %{tags: ^expected_tags} = post_updated} =
+  post_base
+  |> Ecto.Changeset.cast(%{tags: ["new"]}, [:tags])
+  |> Repo.update()
+
+expected_tags = MapSet.new(["cool"])
+
+{:ok, %{tags: ^expected_tags}} =
+  post_updated
+  |> Ecto.Changeset.cast(%{tags: ["cool"]}, [:tags])
+  |> Repo.update()
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+{:ok,
+ %VaxExample.PostWithCrdts{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "post_with_crdts">,
+   content: nil,
+   id: "3567736d-ea8a-48a3-a38a-0ca81bfc6c14",
+   reactions: 0,
+   tags: #MapSet<["cool"]>,
+   title: "Concurrency safe counter and set"
+ }}
+```
+
+## Querying
+
+```elixir
+post =
+  %VaxExample.PostWithCrdts{title: "Reading from id"}
+  |> Repo.insert!()
+
+nil = Repo.get(VaxExample.PostWithCrdts, Ecto.UUID.generate())
+%VaxExample.PostWithCrdts{} = Repo.get(VaxExample.PostWithCrdts, post.id)
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+%VaxExample.PostWithCrdts{
+  __meta__: #Ecto.Schema.Metadata<:loaded, "post_with_crdts">,
+  content: nil,
+  id: "4929a79b-371a-489c-8f03-c2ddd5c594f7",
+  reactions: 0,
+  tags: #MapSet<[]>,
+  title: "Reading from id"
+}
+```
+
+```elixir
+# Simple querying capabilities
+import Ecto.Query
+
+post_a =
+  %VaxExample.PostWithCrdts{title: "Post A"}
+  |> Repo.insert!()
+
+post_b =
+  %VaxExample.PostWithCrdts{title: "Post B"}
+  |> Repo.insert!()
+
+from(VaxExample.PostWithCrdts)
+|> where([p], p.id in ^[post_a.id, post_b.id])
+|> Repo.all()
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+[
+  %VaxExample.PostWithCrdts{
+    __meta__: #Ecto.Schema.Metadata<:loaded, "post_with_crdts">,
+    content: nil,
+    id: "e1bfc996-0330-4990-b492-7e39992d0946",
+    reactions: 0,
+    tags: #MapSet<[]>,
+    title: "Post A"
+  },
+  %VaxExample.PostWithCrdts{
+    __meta__: #Ecto.Schema.Metadata<:loaded, "post_with_crdts">,
+    content: nil,
+    id: "d5a6655e-7a5c-49a2-8daf-52b5a9bcb7c4",
+    reactions: 0,
+    tags: #MapSet<[]>,
+    title: "Post B"
+  }
+]
+```
```

Comment:
Maybe at the end we can flag up some future stuff? I guess that's more a @balegas copy thing.